### ISSUE TITLE
Possible issue with as.mids after 'impute, then transform'

### ISSUE THREAD

**Issue by thomasklausch2** [amices/mice#271]
Dear mice team:

I stumbled over a possible issue with `as.mids`. The issue emerged when I wanted to transform a variable by taking its mean within the imputed data set and then multiply by another variable in the data set. The code below reproduces this setup using the example on p. 172 of van Buuren (2nd ed.). 

```
data = boys[, c("age", "hgt", "wgt", "hc", "reg")]
imp  = mice( data, print = FALSE, seed = 71712)
long1 = complete( imp, 'long', include = T)
pagg = aggregate(wgt ~ .imp, data=long1, FUN = mean)
n = nrow(imp$data)
p    = as.numeric( apply(as.matrix(pagg[,2]), 1, function(x) rep(x, n)) )
long1$phgt = with(long1, p*hgt )
imp.itt = as.mids(long1)
long2 = complete( imp.itt, 'long', include = T)
```

In the example I first take the mean of wgt in each imputed data set and then multiply it by hgt. This is totally arbitrary of course but illustrates the issue.
As expected, `long1$phgt` divided by hgt gives exactly n times the aggregated (mean) values of wgt:

```
> table(long1$phgt/long1$hgt, useNA = 'always')

37.1323810160428 37.1411377005347 37.1411377005348 37.1420601604278 37.1458034759358 37.1482901069519  37.153186827957             <NA> 
             748                3              745              748              748              748              728               20 
```

However, if I then look at the same call after compiling with `as.mids` I find a seemingly non-sensical result. Can you help to clarify if I make a mistake somewhere or if it is indeed an issue in as.mids?

```
> table(long2$phgt/long2$hgt, useNA = 'always')

37.1323810160428 37.1411377005348 37.1420601604278 37.1458034759358 37.1482901069519  37.153186827957             <NA> 
              20               20               20               20               20             4368               20 
```

**Comment by prockenschaub** [comment]
Hi @thomasklausch2 , 

I had a look atthe example that you provide and think I know why you see those differences. In `long1`, all rows with `.imp == 0` correspond to the original data, i.e. they contain the missingness from the `boys` dataset. Then you add a new column `phgt` that calculates those values based on the mean in `wgt` after imputation. Notably this **differs** for each imputation, so the values in `phgt` are different in each imputation. 

The problem now arises when you port the object back into a mids object. What `as.mids` does it treats all rows with  `.imp == 0`. Values that are not missing in these rows will be treated **as if it where originally observed**. All values in rows `.imp %in% 1:5` that are "imputations" of these "observed" rows are thrown away. Only those values are kept that correspond to a missing value in `.imp == 0`, and they are stored in `imp.itt$imp$phgt`. The counts of 20 that you see in your tabulation are the number missing `hgt` values in your original data. These are imputed and when divided by `hgt` result in the corresponding `pagg`s. The large number of values of `37.153186` correspond to the mean in the original data, which is propagated to each imputation. 

Long story short, you have to set the `long1[long1$.imp == 0]$phgt`s in the original data to missing to get your expected result. Code showing this below.

```r
# Load libraries
library(mice)
#> 
#> Attaching package: 'mice'
#> The following objects are masked from 'package:base':
#> 
#>     cbind, rbind
library(tidyr)

# Original example --------------------------------------------------------
data = boys[, c("age", "hgt", "wgt", "hc", "reg")]
imp  = mice( data, print = FALSE, seed = 71712)
long1 = complete( imp, 'long', include = T)
pagg = aggregate(wgt ~ .imp, data=long1, FUN = mean)
n = nrow(imp$data)
p = as.numeric( apply(as.matrix(pagg[,2]), 1, function(x) rep(x, n)) )
long1$phgt = with(long1, p*hgt )
imp.itt = as.mids(long1)
long2 = complete(imp.itt, 'long', include = T)

# Further investigations --------------------------------------------------
wide1 <- spread(long1[, c(".imp", ".id", "phgt")], key = ".imp", value = "phgt")
wide2 <- spread(long2[, c(".imp", ".id", "phgt")], key = ".imp", value = "phgt")

sum(!is.na(data$hgt))  
#> [1] 728
# --> 728 observation have information on hgt

with(wide1, sum(`0` == `1`, na.rm = TRUE))   
#> [1] 0
# --> ALL values of phgt differ between original data (with missings)
#     and imputation 1, because different values of `pagg` were used

with(wide2, sum(`0` == `1`, na.rm = TRUE))
#> [1] 728
# --> values of phgt where hgt was observed in the original data are 
#     the same in imputation 1, because they are treated as if they
#     were originally observed and are thus NOT imputed

# In order to keep all values, set all phgt in the rows corresponding
# to the original data to missing
long1$phgt[1:748] <- NA
imp.itt = as.mids(long1)
long2 = complete(imp.itt, 'long', include = T)
table(long2$phgt/long2$hgt, useNA = 'always')
#> 
#> 37.1323810160428 37.1411377005347 37.1411377005348 37.1420601604278 
#>              748                3              745              748 
#> 37.1458034759358 37.1482901069519             <NA> 
#>              748              748              748
```

**Comment by gerkovink** [comment]
Correct. Or use 
```r 
long1 = complete( imp, 'long', include = T)
```

**Comment by stefvanbuuren** [comment]
> you have to set the long1[long1$.imp == 0]$phgts in the original data to missing

Am I correct in concluding that `as.mids()` cannot recognise and repair this problem?

**Comment by thomasklausch2** [comment]
Thanks to all for finding a work around. I think Stef is correct in his conclusion.

**Comment by prockenschaub** [comment]
Good call, I think it actually kind of can (although not fully automatic). 

As far as I can see, `as.mids()` relies on the characteristics of observations with `.imp == 0` to recreate the `mids` object. For example, it also defaults to the `defaultMethod` even if the original imputation used for example `mice.impute.norm`, simply because this information isn't retained in the long `data.frame`. This means that the solution suggested by @gerkovink of omitting the original data also lead to an error for me:

```r
long1 = complete( imp, 'long', include = FALSE)
...
imp.itt = as.mids(long1)
#> Error in as.mids(long1): Original data not found.
#>  Use `complete(..., action = 'long', include = TRUE)` to save original data.
```

`as.mids` currently accepts a `where` argument, though. This can be used to indicate that this is a column that needs to be imputed: 

```r
long2 = as.mids(long1, where = cbind(imp$where, phgt = TRUE))
table(long2$phgt/long2$hgt, useNA = 'always')
#> 
#> 37.1278999784461 37.1329561757179 37.1477048002915  37.153186827957 37.1607207241353 37.1621883419235             <NA> 
#>              748              748              748              728              748              748               20
```

I am unclear whether this is desirable. `as.mids` seems primarily useful in order to then use the `mids` object with `with`.  If any further imputation where desired, the `methods` arguments needs to be changed manually if it isn't using the defaults. This could be achieved by adding another argument `methods` to `as.mids`. But the same arguemt applies other parameters like `visitSequence`, `post`, etc.

**Comment by gerkovink** [comment]
@stefvanbuuren I believe that @prockenschaub has pointed out a nice and sufficient solution. Fully reproducible and with identical results. 

``` r
library(mice)
#> 
#> Attaching package: 'mice'
#> The following objects are masked from 'package:base':
#> 
#>     cbind, rbind
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(magrittr)

imp <- boys %>% 
  select(age, hgt, wgt, hc, reg) %>%
  mice(print = FALSE, seed = 123)

long1 <- imp %>% 
  mice::complete("long", include = TRUE)

pagg <- aggregate(wgt ~ .imp, data = long1, FUN = mean) %>%
  as.matrix

p <- pagg[, 2] %>% 
  sapply(function(x) rep(x, nrow(imp$data))) %>% 
  as.numeric

long1$phgt <- with(long1, p*hgt)

#as.mids
long2 <- as.mids(long1, where = cbind(imp$where, phgt = TRUE)) %>% 
  complete("long", include = TRUE)

#tables
A <- table(long1$phgt / long1$hgt, useNA = 'always')
B <- table(long2$phgt / long2$hgt, useNA = 'always')
identical(A, B)
#> [1] TRUE
```

<sup>Created on 2020-09-30 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

@thomasklausch2 Sorry for the code changes. This is my process to understand your code.

**Comment by gerkovink** [comment]
I would be hesitant to accommodate `as.mids()` for this scenario - other than @prockenschaub's solution. These values are not bonafide observed data. It'd be nicer to accommodate for these calculations within the algorithm through deterministic imputation of `phgt`. Isn't it better to do this by means of [*passive imputation*](https://www.gerkovink.com/miceVignettes/Passive_Post_processing/Passive_imputation_post_processing.html)

``` r
library(mice)
#> 
#> Attaching package: 'mice'
#> The following objects are masked from 'package:base':
#> 
#>     cbind, rbind
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

set.seed(123)
data <- boys %>% 
  mutate(phgt = NA)
meth <- make.method(data)
pred <- make.predictorMatrix(data)
meth["phgt"] <- "~ I(mean(wgt)*hgt)"
pred[, "phgt"] <- 0
imp <- mice(data, 
            method = meth, 
            predictorMatrix = pred,
            visitSequence = 1:10,
            print = FALSE,
            maxit = 1)
long <- complete(imp, "long", include = FALSE)
table(long$phgt / long$hgt, useNA = 'always')
#> 
#> 37.0891323529412 37.1123943850267 37.1182098930481 37.1182767379679 
#>              748              748              748              748 
#> 37.1577954545455             <NA> 
#>              748                0
```

<sup>Created on 2020-09-30 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by thomasklausch2** [comment]
At this point of the discussion it is perhaps useful for you if I also reveal the background of my question. I attempted to recode a binary (0-1) variable in a way that cases with value 1 would get (1-p) and cases with value 0 get (-p) where p is the proportion of cases with value 1 in the sample. Including this covariate in a regression model would give the intercept the interpretation of the sample (global) mean instead of that of the reference category.

Perhaps there are other ways to achieve such a coding with `mice`, but this was the solution I attempted. Supposedly this can be also reached with passive imputation or even differently.
